Introducing the StableIValue representation of list :D

janeyx99 opened the pull request:
Some important notes:
a) Just like IValues steal the ownership of ArrayRefs and any std::vectors in order to convert the inner elements into IValues, we do the same thing with StableIValue. This O(N) traverse is ineluctable.
b) As a result, since StableIValues are owning and our contract is that to<T>(StableIValue) transfers ownership, you cannot ever convert from StableIValue to a nonowning HeaderOnlyArrayRef<V>.

We handle memory similar to AtenTensorHandle, but we have a StableListHandle!

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #167126
* __->__ #165953
* #165153
* #165152
* #164991

janeyx99 committed:
[draft] look away this commit only has a sample test

[ghstack-poisoned]

janeyx99 committed:
Update on "[draft] look away this commit only has a sample test"




[ghstack-poisoned]

janeyx99 committed:
Update on "[draft] look away this commit only has a sample test"




[ghstack-poisoned]

janeyx99 committed:
Update on "[draft] look away this commit only has a sample test"




[ghstack-poisoned]

janeyx99 committed:
Update on "Introducing the StableIValue representation of list :D"




[ghstack-poisoned]

janeyx99 committed:
Update on "Introducing the StableIValue representation of list :D"




[ghstack-poisoned]

swolchok commented on torch/csrc/stable/c/shim.h:
```diff
@@ -37,6 +37,34 @@ AOTI_TORCH_EXPORT AOTITorchError torch_library_impl(
+torch_new_list_reserve_size(size_t size, StableListHandle* ret);
+
+// new list handle steals ownership of underlying elements from orig_handle
+AOTI_TORCH_EXPORT AOTITorchError torch_new_list_handle(
```
maybe call it torch_move_list_handle? also this seems like an odd function to need; why can't the caller just continue to use `orig_handle`?

janeyx99 replied:
yea this was leftover when i had an overcomplicated design for memory, but we do not need it anymore.

swolchok commented on torch/csrc/stable/c/shim.h:
```diff
@@ -37,6 +37,34 @@ AOTI_TORCH_EXPORT AOTITorchError torch_library_impl(
+
+// deletes the underlying list referenced by list_handle
+AOTI_TORCH_EXPORT AOTITorchError
+torch_delete_list_object(StableListHandle list_handle);
```
why does this function have "object" in the name when none of the other list functions do?

swolchok commented on torch/csrc/stable/c/shim.h:
```diff
@@ -37,6 +37,34 @@ AOTI_TORCH_EXPORT AOTITorchError torch_library_impl(
+AOTI_TORCH_EXPORT AOTITorchError torch_list_get(
+    StableListHandle list_handle,
+    size_t index,
+    StableIValue* element);
```
torch_list_get_item?

janeyx99 replied:
But this one should be `torch_list_at(...)`

swolchok reviewed (commented):
I had a lot of confusion around the ownership model before I went back and reviewed how ownership of tensors works and confirmed that (I think) StableIValue owns lists to the same extent that it owns tensors -- it takes care of its own cleanup. I think it's probably OK but might be something to think about.

janeyx99 committed:
Update on "Introducing the StableIValue representation of list :D"

Some important notes:
a) Just like IValues steal the ownership of ArrayRefs and any std::vectors in order to convert the inner elements into IValues, we do the same thing with StableIValue. This O(N) traverse is ineluctable.
b) As a result, since StableIValues are owning and our contract is that to<T>(StableIValue) transfers ownership, you cannot ever convert from StableIValue to a nonowning HeaderOnlyArrayRef<V>.

We handle memory similar to AtenTensorHandle, but we have a StableListHandle!




[ghstack-poisoned]

janeyx99 commented on torch/csrc/stable/c/shim.h:
```diff
@@ -37,6 +37,34 @@ AOTI_TORCH_EXPORT AOTITorchError torch_library_impl(
+    StableListHandle* new_handle);
+
+AOTI_TORCH_EXPORT AOTITorchError
+torch_list_size(StableListHandle list_handle, size_t* size);
```
I'm following the semantic where if it's blah->api(), I'm naming it torch_blah_api, so this one stays.

malfet replied:
Hmm, on one hand I understand the desire to keep `error_t foo(inp, *out)` semantic, but on the other hand, the only it it could fail if `list_handle` is nullptr, isn't it?  Should we indeed for the sake of speed change semantic a bit to `size_t torch_list_get_size(StableListHandle)`?

janeyx99 committed:
Update on "Introducing the StableIValue representation of list :D"

Some important notes:
a) Just like IValues steal the ownership of ArrayRefs and any std::vectors in order to convert the inner elements into IValues, we do the same thing with StableIValue. This O(N) traverse is ineluctable.
b) As a result, since StableIValues are owning and our contract is that to<T>(StableIValue) transfers ownership, you cannot ever convert from StableIValue to a nonowning HeaderOnlyArrayRef<V>.

We handle memory similar to AtenTensorHandle, but we have a StableListHandle!




[ghstack-poisoned]

swolchok commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/csrc/kernel.cpp:
```diff
@@ -478,6 +478,62 @@ STABLE_TORCH_LIBRARY_IMPL(libtorch_agnostic, CompositeExplicitAutograd, m) {
+  const auto num_args = 2;
+  StableIValue stack[num_args];
+  stack[0] = from(self);
+  stack[1] = from(other);
```
nit: this is acceptable, but since num_args isn't actually used anywhere else and I hate being nagged by linters, I would write

std::array<StableIValue, 2> stack = {{from(self), from(other)}};

swolchok commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/csrc/kernel.cpp:
```diff
@@ -478,6 +478,62 @@ STABLE_TORCH_LIBRARY_IMPL(libtorch_agnostic, CompositeExplicitAutograd, m) {
+}
+
+void my__foreach_mul_(torch::headeronly::HeaderOnlyArrayRef<Tensor> self, torch::headeronly::HeaderOnlyArrayRef<Tensor> other) {
+  const auto num_args = 2;
```
ditto

malfet commented on torch/csrc/stable/c/shim.h:
```diff
@@ -37,6 +37,29 @@ AOTI_TORCH_EXPORT AOTITorchError torch_library_impl(
+AOTI_TORCH_EXPORT AOTITorchError torch_list_at(
+    StableListHandle list_handle,
+    size_t index,
+    StableIValue* element);
```
Same as above: if  `at::List::operator[]` doesn't do boundary checks, this API shouldn't do them either

janeyx99 commented on torch/csrc/stable/stableivalue_conversions.h:
```diff
@@ -342,6 +374,33 @@ struct ToImpl<torch::stable::Tensor> {
+    TORCH_ERROR_CODE_CHECK(torch_list_size(list_handle, &size));
+    std::vector<T> result;
+    result.reserve(size);
+    for (size_t i = 0; i < size; i++) {
```
We'd have to move irange to headeronly first, but otherwise we can

swolchok commented on torch/csrc/stable/c/shim.h:
```diff
@@ -37,6 +37,29 @@ AOTI_TORCH_EXPORT AOTITorchError torch_library_impl(
+    StableIValue* element);
+
+AOTI_TORCH_EXPORT AOTITorchError
+torch_list_push_back(StableListHandle list_handle, StableIValue element);
```
is it permissible to be wrong about the reserved size?

janeyx99 replied:
I think reserving size is just a performance optimization, it doesn't actually populate the tensor. So after reserving the size is still 0.

swolchok commented on torch/csrc/stable/c/shim.h:
```diff
@@ -37,6 +37,29 @@ AOTI_TORCH_EXPORT AOTITorchError torch_library_impl(
+// returns an owning reference of a StableList. callee is responsible for
+// freeing memory.
+AOTI_TORCH_EXPORT AOTITorchError
+torch_new_list_reserve_size(size_t size, StableListHandle* ret);
```
interestingly, this does not correspond to any constructor of std::vector, though I like avoiding a second crossing of the ABI boundary just to do a reserve.

swolchok commented on torch/csrc/stable/stableivalue_conversions.h:
```diff
@@ -342,6 +374,33 @@ struct ToImpl<torch::stable::Tensor> {
+// std::vector<T> should be represented as a StableListHandle
+// filled with StableIValues
+// The new std::vector steals ownership of the underlying elements
+// and StableListHandle is freed.
```
this isn't quite right -- the list the handle refers to is freed, not the handle itself, right?

janeyx99 commented on torch/csrc/stable/c/shim.h:
```diff
@@ -37,6 +37,29 @@ AOTI_TORCH_EXPORT AOTITorchError torch_library_impl(
+AOTI_TORCH_EXPORT AOTITorchError torch_list_at(
+    StableListHandle list_handle,
+    size_t index,
+    StableIValue* element);
```
Oh I see....this function indeed doesn't have the same semantic as the C++ API as there's no way to set with my C API. I'll rename and add a way to set.

I agree we should not rush to add things

janeyx99 committed:
Update on "Introducing the StableIValue representation of list :D"

Some important notes:
a) Just like IValues steal the ownership of ArrayRefs and any std::vectors in order to convert the inner elements into IValues, we do the same thing with StableIValue. This O(N) traverse is ineluctable.
b) As a result, since StableIValues are owning and our contract is that to<T>(StableIValue) transfers ownership, you cannot ever convert from StableIValue to a nonowning HeaderOnlyArrayRef<V>.

We handle memory similar to AtenTensorHandle, but we have a StableListHandle!




[ghstack-poisoned]

swolchok commented on torch/csrc/stable/stableivalue_conversions.h:
```diff
@@ -192,6 +193,37 @@ struct FromImpl<torch::stable::Tensor> {
+        torch_new_list_reserve_size(val.size(), &new_list_handle));
+    for (const auto& elem : val) {
+      TORCH_ERROR_CODE_CHECK(torch_list_push_back(new_list_handle, from(elem)));
+    }
```
you're going to leak memory for the list and for previously-pushed elements if this TORCH_ERROR_CODE_CHECK ever fails, courtesy of StableIValue not having a destructor

swolchok commented on torch/csrc/stable/stableivalue_conversions.h:
```diff
@@ -342,6 +374,33 @@ struct ToImpl<torch::stable::Tensor> {
+    result.reserve(size);
+    for (size_t i = 0; i < size; i++) {
+      StableIValue element;
+      TORCH_ERROR_CODE_CHECK(torch_list_get_item(list_handle, i, &element));
```
ditto leaky

swolchok reviewed (commented):
looking pretty good

janeyx99 committed:
Update on "Introducing the StableIValue representation of list :D"

Some important notes:
a) Just like IValues steal the ownership of ArrayRefs and any std::vectors in order to convert the inner elements into IValues, we do the same thing with StableIValue. This O(N) traverse is ineluctable.
b) As a result, since StableIValues are owning and our contract is that to<T>(StableIValue) transfers ownership, you cannot ever convert from StableIValue to a nonowning HeaderOnlyArrayRef<V>.

We handle memory similar to AtenTensorHandle, but we have a StableListHandle!




[ghstack-poisoned]

swolchok commented on torch/csrc/stable/stableivalue_conversions.h:
```diff
@@ -202,12 +202,21 @@ struct FromImpl<torch::headeronly::HeaderOnlyArrayRef<T>> {
+            torch_list_push_back(new_list_handle, from(elem)));
+      }
+      return from(new_list_handle);
+    } catch (const std::runtime_error& e) {
```
don't you need to catch all flavors of exception? why is it ok to leak on things that aren't runtime_error?

swolchok commented on torch/csrc/stable/stableivalue_conversions.h:
```diff
@@ -387,17 +396,22 @@ struct ToImpl<std::vector<T>> {
+    } catch (const std::runtime_error& e) {
+      // clean up memory if an exception is thrown, and rethrow
+      TORCH_ERROR_CODE_CHECK(torch_delete_list(list_handle));
+      throw e;
```
same as above

janeyx99 commented on torch/csrc/stable/stableivalue_conversions.h:
```diff
@@ -202,12 +202,21 @@ struct FromImpl<torch::headeronly::HeaderOnlyArrayRef<T>> {
+        // clean up memory if an error was thrown
+        TORCH_ERROR_CODE_CHECK(torch_delete_list(new_list_handle));
+      }
+      throw e;
```
Wait does that do the same thing? Does that rethrow the error?

janeyx99 replied:
Ohhh it issss

swolchok replied:
generally rethrowing can also do stuff like cause backtraces to be preserved.

malfet replied:
`throw e;` can cause object slicing, while `throw` preserve type of original exception?

janeyx99 committed:
Update on "Introducing the StableIValue representation of list :D"

Some important notes:
a) Just like IValues steal the ownership of ArrayRefs and any std::vectors in order to convert the inner elements into IValues, we do the same thing with StableIValue. This O(N) traverse is ineluctable.
b) As a result, since StableIValues are owning and our contract is that to<T>(StableIValue) transfers ownership, you cannot ever convert from StableIValue to a nonowning HeaderOnlyArrayRef<V>.

We handle memory similar to AtenTensorHandle, but we have a StableListHandle!




[ghstack-poisoned]

swolchok commented on torch/csrc/shim_common.cpp:
```diff
@@ -4,12 +4,55 @@
+}
+
+AOTI_TORCH_EXPORT AOTITorchError
+torch_list_size(StableListHandle list_handle, size_t* size) {
```
FWIW, it's harmless if size() gets inlined and so I am weakly in favor of it because it makes everything visually uniform (lack of `AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE` is known-bad). https://godbolt.org/z/zbqbhez87
